src/lexparse/parser.go

```diff
@@ -0,0 +1,218 @@
+package lexparse
+
+import (
+  "fmt"
+  "strings"
+)
+
+type NodeType int
+
+const (
+  DeclarationS NodeType = iota
+  ExpressionS
+  TypeDefnS
+
+  ArrayTypeN
+  SetTypeN
+  MapTypeN
+  FunctionTypeN
+
+  DeclaredChainN
+  DeclaredN
+  DeclaredLeftN
+  FunctionArgdefN
+
+  SimpleExprN
+)
+
+var simpleExprTokenTypes map[tokenType]bool = map[tokenType]bool {
+  SymbolT: true,
+  StringT: true,
+  CharT: true,
+  IntT: true,
+  FloatT: true,
+  BoolT: true,
+  NullT: true,
+}
+
+// How the Node struct is used:
+//
+// token is the symbol token/int token/etc for SimpleExprN,
+// and the binop/comparison/setter/prefix/suffix token for those
+// nodetypes. Otherwise it's usually just the first token of
+// a node and not important aside for parse error printouts.
+//
+// left and right are obviously used in branching, but some nodetypes
+// only have one descendant, e.g. ExpressionS, PrefixN, SuffixN.
+// For node types with one descendant, only left is used.
+// SimpleExprN is the only node type that uses neither left or right.
+
+type Node struct {
```

Comment:
Another way to do this is to have a bunch of separate Node types which implement a common interface, e.g. https://github.com/iafisher/torino/blob/master/parser/ast.go

Reply:
Ah, I wasn't aware that Go switch-case statements could work [like this](https://github.com/iafisher/torino/blob/113b228ecf5b3c3ced78722b42e04f90e11d6810/compiler/compiler.go#L35) - that could be handy.

It does look like your way involves a fair amount of boilerplate code... what would you say are its advantages?

Reply:
Yeah, the boilerplate is an unfortunate shortcoming of Go. I'd say the main advantages are that you get semantic types like `ForNode` and `WhileNode` instead of everything being the same `Node` type, and consequently your individual nodes don't have fields that they don't use.

Reply:
Gotcha. I don't think I'll refactor now but there's a good chance of me doing so later.